### ISSUE TITLE
[Bug]: App crashes intercepting Chrome with "Cannot read properties of null (reading 'finishShutdown')" (code 1)

### ISSUE THREAD

**Issue by fdciabdul** [httptoolkit/httptoolkit#439]
### Repro steps

1 . Open the App
2. Click Chrome
3. Visit some web
4.  Crashed

### How often does this bug happen?

Every time

### The desktop OS you're using

Ubtuntu 22

### Details of other apps/devices

- Intercepting chrome 114.0.5735.133

### Error screenshot

![image](https://github.com/httptoolkit/httptoolkit/assets/31664438/ca667959-b495-4b92-b74a-f002fbabdfb4)

**Comment by pimterry** [comment]
I see, thanks for reporting this. Can you share your logs? You can find them in the menu under Help -> View your HTTP Toolkit logs.

They'll reset when the app restarts, which is a bit awkward for crashing cases like this, but you can grab them by starting the app, opening the logs folder immediately, reproducing the crash, and then sending the log file from that folder. That should contain all the details of what's going wrong here.

**Comment by pimterry** [comment]
Oh, also can you share any more details about the website that triggers this,  and exactly what's happening at that time? E.g. are you loading a page, closing a page, clicking something, doing nothing at all.

On my machine I've tested on Ubuntu with the same Chrome version, and I can't reproduce this unfortunately. I suspect it is related to some specific client or server behaviour that happens in certain cases, so anything you can share about exactly what is going on at the moment the app crashes would be super helpful.

**Comment by fdciabdul** [comment]
@pimterry 

i don't know where is view logs menu 

![image](https://github.com/httptoolkit/httptoolkit/assets/31664438/8734b3db-a70b-49d5-8efb-065cb5168937)


here's the video about it 


https://github.com/httptoolkit/httptoolkit/assets/31664438/450aca40-3934-4b67-9778-42d1742864b6

**Comment by fdciabdul** [comment]
> Oh, also can you share any more details about the website that triggers this, and exactly what's happening at that time? E.g. are you loading a page, closing a page, clicking something, doing nothing at all.
> 
> On my machine I've tested on Ubuntu with the same Chrome version, and I can't reproduce this unfortunately. I suspect it is related to some specific client or server behaviour that happens in certain cases, so anything you can share about exactly what is going on at the moment the app crashes would be super helpful.

almost any website, it happen randomly, sometimes it works withint 2-5 minutes then HTTPToolkit crashed , then the intercept log was cleared , but sometimes it happen quickly like i just open the web and the app was crashed again

**Comment by pimterry** [comment]
> i don't know where is view logs menu

Err, depends on your window manager, but wherever the normal application File / Edit / View / etc menu lives. Pressing `alt` normally jumps to it.

Alternatively you can just open them directly though - on Linux the logs from the last run are stored at `/home/<username>/.config/httptoolkit/logs/last-run.log`.

**Comment by fdciabdul** [comment]
> > i don't know where is view logs menu
> 
> Err, depends on your window manager, but wherever the normal application File / Edit / View / etc menu lives. Pressing `alt` normally jumps to it.
> 
> Alternatively you can just open them directly though - on Linux the logs from the last run are stored at `/home/<username>/.config/httptoolkit/logs/last-run.log`.

here the logs 

```bash

--- Launching HTTP Toolkit desktop v1.12.7 ---
INFO: Initialising UI (version 1daef1176df850437cb1c4942e5b47ebb4056c8e)
INFO: Account store initialized
INFO: UI store initialized
INFO: Proxy settings loaded
INFO: Server initialization failed TypeError: Failed to fetch
INFO: API store initialized
INFO: Server initialization failed TypeError: Failed to fetch
INFO: Server initialization failed TypeError: Failed to fetch
(node:255229) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Config checked in 34 ms
Certificates setup in 16 ms
Standalone server started in 10 ms
Server started in 27 ms
Total startup took 87 ms
(node:255229) Warning: Using insecure HTTP parsing
Mock session started, http on port 8000, webrtc enabled
Docker not available: connect ENOENT /var/run/docker.sock
INFO: Server started
Webextension installed at /tmp/httptoolkit-webextension
INFO: Config loaded
INFO: Server started on port 8000
INFO: Proxy store initialized (server version 1.12.8)
INFO: Rules store initialized
INFO: Events store initialized
INFO: Interceptor store initialized
INFO: App started, rendering
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
Browser cache updated
INFO: Service worker loaded
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
Firefox started
Request to /check-cert received
ATTENTION: default value of option mesa_glthread overridden by environment.
ATTENTION: default value of option mesa_glthread overridden by environment.
ATTENTION: default value of option mesa_glthread overridden by environment.
Request to /test-https successfully received
Failed to handle request: socket hang up
INFO: Previous server version was 1.12.8
ERROR: ResizeObserver loop limit exceeded
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
Failed to handle request: socket hang up
Failed to handle request: socket hang up
Failed to handle request: socket hang up
Failed to handle request: socket hang up
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
INFO: Reporting error: TypeError: Failed to fetch [object Object]
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
Shutting down after API call...
Uninstalling webextension from /tmp/httptoolkit-webextension
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Exiting due to channel error.
Firefox closed

```

**Comment by pimterry** [comment]
Thanks! Unfortunately though there's no crash in here - it looks like this was shut down cleanly (`Shutting down after API call...` means you closed the app on purpose). Can you reproduce the crash, and then share the logs from that?

**Comment by fdciabdul** [comment]
> Thanks! Unfortunately though there's no crash in here - it looks like this was shut down cleanly (`Shutting down after API call...` means you closed the app on purpose). Can you reproduce the crash, and then share the logs from that?

i do .. here's the log after crashed 

```bash
--- Launching HTTP Toolkit desktop v1.12.7 ---
INFO: Initialising UI (version 1daef1176df850437cb1c4942e5b47ebb4056c8e)
INFO: Account store initialized
INFO: UI store initialized
INFO: Proxy settings loaded
INFO: Server initialization failed TypeError: Failed to fetch
INFO: API store initialized
INFO: Server initialization failed TypeError: Failed to fetch
INFO: Server initialization failed TypeError: Failed to fetch
(node:296329) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Config checked in 18 ms
Certificates setup in 7 ms
Standalone server started in 6 ms
Server started in 18 ms
Total startup took 49 ms
(node:296329) Warning: Using insecure HTTP parsing
Mock session started, http on port 8000, webrtc enabled
Docker not available: connect ENOENT /var/run/docker.sock
INFO: Server started
Webextension installed at /tmp/httptoolkit-webextension
INFO: Config loaded
INFO: Server started on port 8000
INFO: Proxy store initialized (server version 1.12.8)
INFO: Rules store initialized
INFO: Events store initialized
INFO: Interceptor store initialized
INFO: App started, rendering
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
[296456:296456:0622/191236.694710:ERROR:interface_endpoint_client.cc(697)] Message 0 rejected by interface blink.mojom.WidgetHost
Browser cache updated
INFO: Service worker loaded
httptoolkit-server: Updating CLI... fetching manifest
httptoolkit-server: Updating CLI... already on latest version: 1.12.8
TypeError: Cannot read properties of null (reading 'finishShutdown')
    at JSStreamSocket.finishShutdown (node:internal/js_stream_socket:160:12)
    at node:internal/js_stream_socket:147:14
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21)

```

**Comment by fdciabdul** [comment]
also , if error happen , then if i'm ignore it 

the app was freezing 

can't even close it 


https://github.com/httptoolkit/httptoolkit/assets/31664438/cf6e0e5b-d1d2-44d6-994e-c680ac16f5dc

**Comment by pimterry** [comment]
Thanks, that `TypeError` at the end is what I was looking for. I think this is likely to be caused by a Node.js bug, as the stack trace is deep within their networking code, and I've just updated to a new release there very recently.

I'll talk to the Node team about possible fixes here. In the meantime, I've just rolled back that update - next time you run HTTP Toolkit, it will update in the background to pick up this fix (takes a minute or so) and then the subsequent run this should be resolved. You can check by hovering over the HTTP logo to see the internal component versions currently running - the fixed version is server v1.12.9.

**Comment by fdciabdul** [comment]
> Thanks, that `TypeError` at the end is what I was looking for. I think this is likely to be caused by a Node.js bug, as the stack trace is deep within their networking code, and I've just updated to a new release there very recently.
> 
> I'll talk to the Node team about possible fixes here. In the meantime, I've just rolled back that update - next time you run HTTP Toolkit, it will update in the background to pick up this fix (takes a minute or so) and then the subsequent run this should be resolved. You can check by hovering over the HTTP logo to see the internal component versions currently running - the fixed version is server v1.12.9.


Thank you so much for your swift response and hard work, @pimterry ! I truly appreciate you taking the time to address this bug and I'm pleased to hear there's a fix for this issue. I'll make sure to check the server version when I run HTTP Toolkit again.

Just a small suggestion from me, it might be really helpful if there were specific notifications or messages when updates are available. That way, we could have more control over when and what gets updated, which might help prevent issues like this one in the future. Thanks again for all your hard work!

**Comment by pimterry** [comment]
Just to update here - the Node version has now been rolled back, and the corresponding Node bug is tracked here: https://github.com/nodejs/node/issues/48519. If you run HTTP Toolkit it'll update in the background (takes a minute or so) and then the following run this should be working correctly.

I'm going to close this as fixed now, but do let me know if you're still having issues after that update, or if this comes back in future!
